Index repair

gguttikonda opened the pull request:
-----
PR for https://github.com/JanusGraph/janusgraph/issues/1895

Thank you for contributing to JanusGraph!

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there an issue associated with this PR? Is it referenced in the commit message?
- [ ] Does your PR body contain #xyz where xyz is the issue number you are trying to resolve?
- [x] Has your PR been rebased against the latest commit within the target branch (typically `master`)?
- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [x] Have you written and/or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](https://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE.txt file, including the main LICENSE.txt file in the root of this repository?
- [ ] If applicable, have you updated the NOTICE.txt file, including the main NOTICE.txt file found in the root of this repository?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?
- [ ] If this PR is a documentation-only change, have you added a `[doc only]`
  tag to the first line of your commit message to avoid spending CPU cycles in
  Travis CI when no code, tests, or build configuration are modified?

### Note:
Please ensure that once the PR is submitted, you check Travis CI for build issues and submit an update to your PR as soon as possible.

gguttikonda commented:
This is PR for https://github.com/JanusGraph/janusgraph/issues/1895

FlorianHockmann commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 1
           

Complexity increasing per file
==============================
- janusgraph-core/src/main/java/org/janusgraph/graphdb/olap/job/IndexRepairJob.java  1
- janusgraph-core/src/test/java/org/janusgraph/graphdb/database/IndexSerializerTest.java  1
         

Clones added
============
- janusgraph-core/src/test/java/org/janusgraph/graphdb/database/IndexSerializerTest.java  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/janusgraph/janusgraph/pullRequest?prid=4609964&bid=15465456)

FlorianHockmann commented on janusgraph-cql/src/main/java/org/janusgraph/diskstorage/cql/CQLKeyColumnValueStore.java:
```diff
@@ -100,6 +103,7 @@
  * An implementation of {@link KeyColumnValueStore} which stores the data in a CQL connected backend.
  */
 public class CQLKeyColumnValueStore implements KeyColumnValueStore {
+    private static final Logger log = LoggerFactory.getLogger(CQLKeyColumnValueStore.class);
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Avoid unused private fields such as 'log'.](https://app.codacy.com/app/JanusGraph/janusgraph/pullRequest?prid=4609964)

FlorianHockmann commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- janusgraph-core/src/main/java/org/janusgraph/graphdb/olap/job/IndexRepairJob.java  1
- janusgraph-core/src/test/java/org/janusgraph/graphdb/database/IndexSerializerTest.java  1
         

Clones added
============
- janusgraph-core/src/test/java/org/janusgraph/graphdb/database/IndexSerializerTest.java  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/janusgraph/janusgraph/pullRequest?prid=4609964&bid=15465456)

porunov commented:
@gguttikonda Would you be able to sign EasyCLA and fix failed tests in travis?

https://travis-ci.org/JanusGraph/janusgraph/builds/620314346?utm_source=github_status&utm_medium=notification

gani8780 commented:
> @gguttikonda Would you be able to sign EasyCLA and fix failed tests in travis?
> 
> https://travis-ci.org/JanusGraph/janusgraph/builds/620314346?utm_source=github_status&utm_medium=notification


Thanks @porunov. I have accepted EastCLA. I'll work on travis build failures and keep you posted.

gguttikonda commented:
Hi @porunov , Test failures are now fixed. Thank you for advising me on this. It helped me to change the approach a little. Below is a summary of change.

* Reverted isExhausted method in KeyIterator
* Removed the retryCount check in StandardScannerExecutor.java
    * Old
        `while (!pullThreads[i].isFinished() && retryCount < TIMEOUT_MS / TIME_PER_TRY && qr == null) { ...`     
    * New
`while (!pullThreads[i].isFinished() && qr == null) { ...`
    * Moved the `finished = true` code in DataPuller into finally block. Making sure for failure conditions or exceptions are setting the thread as finished.

farodin91 commented on janusgraph-cassandra/src/main/java/org/janusgraph/diskstorage/cassandra/astyanax/AstyanaxKeyColumnValueStore.java:
```diff
@@ -432,6 +432,7 @@ public void remove() {
             };
         }
 
+                
```
Could you remove all independent file changes?

farodin91 replied:
From the pr

gguttikonda replied:
Hi @farodin91 , Thank you for the feedback. I checked in my changes with your review feedback.

FlorianHockmann commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 1
           

Complexity increasing per file
==============================
- janusgraph-core/src/test/java/org/janusgraph/graphdb/database/IndexSerializerTest.java  1
- janusgraph-core/src/main/java/org/janusgraph/graphdb/olap/job/IndexRepairJob.java  1
         

Complexity decreasing per file
==============================
+ janusgraph-core/src/main/java/org/janusgraph/diskstorage/keycolumnvalue/scan/StandardScannerExecutor.java  -1
         

Clones added
============
- janusgraph-core/src/test/java/org/janusgraph/graphdb/database/IndexSerializerTest.java  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/JanusGraph/janusgraph/pullRequest?prid=4609964&bid=15465456)

FlorianHockmann commented on janusgraph-cql/src/test/java/org/janusgraph/diskstorage/cql/CQLResultSetKeyIteratorTest.java:
```diff
@@ -16,6 +16,7 @@
 
 import static org.junit.jupiter.api.Assertions.assertEquals;
 import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertTrue;
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Avoid unused imports such as 'org.junit.jupiter.api.Assertions.assertTrue'](https://app.codacy.com/manual/JanusGraph/janusgraph/pullRequest?prid=4609964)

farodin91 commented on janusgraph-berkeleyje/src/test/resources/log4j.properties:
```diff
@@ -16,12 +16,12 @@ log4j.appender.A1=org.apache.log4j.FileAppender
 log4j.appender.A1.File=target/test.log
 log4j.appender.A1.Threshold=ALL
 log4j.appender.A1.layout=org.apache.log4j.PatternLayout
-log4j.appender.A1.layout.ConversionPattern=%-4r [%t] %-5p %c %x - %m%n
```
Why did you change the log pattern of Berkeleyje?

gguttikonda replied:
I apologize. I was trying to isolate some test failures locally and was trying to find the timestamps for each test. I reverted this change now.

farodin91 commented on janusgraph-cassandra/src/main/java/org/janusgraph/diskstorage/cassandra/embedded/CassandraEmbeddedKeyColumnValueStore.java:
```diff
@@ -335,6 +335,7 @@ public RowIterator(Token minimum, Token maximum, SliceQuery sliceQuery, int page
             this.nowMillis = times.getTime().toEpochMilli();
             this.keys = getRowsIterator(getKeySlice(minimum, maximum, sliceQuery, pageSize, nowMillis));
         }
+        
```
nit: extra line

gguttikonda replied:
Removed :)

farodin91 commented on janusgraph-core/src/test/resources/log4j.properties:
```diff
@@ -0,0 +1,32 @@
+# Copyright 2019 JanusGraph Authors
```
Could you delete this file?

gguttikonda replied:
Sorry about this one! I deleted this file.

farodin91 commented on janusgraph-cql/src/main/java/org/janusgraph/diskstorage/cql/CQLKeyColumnValueStore.java:
```diff
@@ -426,8 +426,9 @@ public KeyIterator getKeys(final SliceQuery query, final StoreTransaction txh) t
         private int index;
         private int paginatedResultSize;
         private final Supplier<Statement> statementSupplier;
+        private AtomicLong recordCount = new AtomicLong();
```
Why did you done these changes?

gguttikonda replied:
I was trying to keep track of how many CQL records being read from DB and trying to keep a custom metric for the same. I  forgot removing counter after I removed that part the code.

FlorianHockmann commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 1
           

Complexity increasing per file
==============================
- janusgraph-core/src/main/java/org/janusgraph/graphdb/olap/job/IndexRepairJob.java  1
- janusgraph-test/src/test/java/org/janusgraph/graphdb/database/IndexSerializerTest.java  1
         

Complexity decreasing per file
==============================
+ janusgraph-core/src/main/java/org/janusgraph/diskstorage/keycolumnvalue/scan/StandardScannerExecutor.java  -1
         

Clones added
============
- janusgraph-test/src/test/java/org/janusgraph/graphdb/database/IndexSerializerTest.java  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/JanusGraph/janusgraph/pullRequest?prid=4609964&bid=15465456)

FlorianHockmann commented on janusgraph-cql/src/main/java/org/janusgraph/diskstorage/cql/CQLKeyColumnValueStore.java:
```diff
@@ -51,10 +51,10 @@
 import java.util.List;
 import java.util.Map;
 import java.util.concurrent.ExecutorService;
+import java.util.concurrent.atomic.AtomicLong;
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Avoid unused imports such as 'java.util.concurrent.atomic.AtomicLong'](https://app.codacy.com/manual/JanusGraph/janusgraph/pullRequest?prid=4609964)

FlorianHockmann commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- janusgraph-core/src/main/java/org/janusgraph/graphdb/olap/job/IndexRepairJob.java  1
- janusgraph-test/src/test/java/org/janusgraph/graphdb/database/IndexSerializerTest.java  1
         

Complexity decreasing per file
==============================
+ janusgraph-core/src/main/java/org/janusgraph/diskstorage/keycolumnvalue/scan/StandardScannerExecutor.java  -1
         

Clones added
============
- janusgraph-test/src/test/java/org/janusgraph/graphdb/database/IndexSerializerTest.java  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/JanusGraph/janusgraph/pullRequest?prid=4609964&bid=15465456)

FlorianHockmann commented on janusgraph-cql/src/main/java/org/janusgraph/diskstorage/cql/CQLKeyColumnValueStore.java:
```diff
@@ -448,15 +447,17 @@ public Row next() {
-            lastPagingState = currentResultSet.getExecutionInfo().getPagingState();
+            // Using getPagingStateUnsafe rather than getPagingState, to avoid the md5
+            // compute of the pageState, which can be quite expensive.
+            lastPagingState = currentResultSet.getExecutionInfo().getPagingStateUnsafe();
```
How is this related to the index improvements? If it's completely unrelated, then please create a separate PR (and probably also its own issue) for this.

gguttikonda replied:
It was kind of related. We are using CQL adapter, and we found that the above change is one of the biggest performance improvements while indexing. Before the above change, it was taking ~4-5 hours for ~1M vertexes and after the change, it's about 15mins. The largest time being spent was on `getPagingState` which does md5sum compute of the entire resultset.

FlorianHockmann replied:
That is of course a significant improvement, but I would still prefer to handle that outside of this PR as it's a CQL specific improvement and such a significant improvement shouldn't be "hidden" in this PR.

gguttikonda replied:
Sure @FlorianHockmann. I reverted the change for CQLKeyColumnValueStore, and will create a new PR only for CQL optimization.

FlorianHockmann commented on janusgraph-core/src/main/java/org/janusgraph/graphdb/olap/job/IndexUpdateJob.java:
```diff
@@ -139,6 +140,10 @@ public void workerIterationEnd(ScanMetrics metrics) {
             metrics.incrementCustom(FAILED_TX);
             throw e;
+        } finally {
+            log.info("Index {} metrics success-tx {} doc-updates  {} success {}", indexName,
```
(nitpick) I would change the format slightly, e.g., like this:
```
"Index {} metrics: success-tx: {} doc-updates: {} succeeded: {}"
```

gguttikonda replied:
Updated :-)

FlorianHockmann commented on janusgraph-core/src/main/java/org/janusgraph/graphdb/database/IndexSerializer.java:
```diff
@@ -355,8 +356,10 @@ public void reindexElement(JanusGraphElement element, MixedIndexType index, Map<
                 element.values(key.name()).forEachRemaining(value->entries.add(new IndexEntry(key2Field(field), value)));
             }
         }
-        final Map<String, List<IndexEntry>> documents = documentsPerStore.computeIfAbsent(index.getStoreName(), k -> Maps.newHashMap());
```
I'm not really familiar with this part of the code base, but could you quickly explain why this isn't necessary any more?

gguttikonda replied:
Hi @FlorianHockmann,
This code is removed for 2 reasons.
1.  `getDocuments(documentsPerStore, index)` does exact same thing, and hence it was duplicate execution of `computeIfAbsent`
2. `documents` variable was not being used.

FlorianHockmann replied:
Thanks for the explanation! That makes sense of course.

FlorianHockmann commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 1
           

Complexity increasing per file
==============================
- janusgraph-core/src/main/java/org/janusgraph/graphdb/olap/job/IndexRepairJob.java  1
- janusgraph-test/src/test/java/org/janusgraph/graphdb/database/IndexSerializerTest.java  2
         

Complexity decreasing per file
==============================
+ janusgraph-core/src/main/java/org/janusgraph/diskstorage/keycolumnvalue/scan/StandardScannerExecutor.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/JanusGraph/janusgraph/pullRequest?prid=4609964&bid=15465456)

FlorianHockmann commented on janusgraph-test/src/test/java/org/janusgraph/graphdb/database/IndexSerializerTest.java:
```diff
@@ -0,0 +1,143 @@
+
+        doReturn(false).when(mit).hasSchemaTypeConstraint();
+
+        Map<String, Map<String, List<IndexEntry>>> docStore = new HashMap<>();
```
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Avoid unused local variables such as 'docStore'.](https://app.codacy.com/manual/JanusGraph/janusgraph/pullRequest?prid=4609964)

FlorianHockmann commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- janusgraph-test/src/test/java/org/janusgraph/graphdb/database/IndexSerializerTest.java  2
- janusgraph-core/src/main/java/org/janusgraph/graphdb/olap/job/IndexRepairJob.java  1
         

Complexity decreasing per file
==============================
+ janusgraph-core/src/main/java/org/janusgraph/diskstorage/keycolumnvalue/scan/StandardScannerExecutor.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/JanusGraph/janusgraph/pullRequest?prid=4609964&bid=15465456)

gguttikonda commented on janusgraph-test/src/test/java/org/janusgraph/graphdb/database/IndexSerializerTest.java:
```diff
@@ -0,0 +1,152 @@
+        StandardJanusGraphTx tx = mock(StandardJanusGraphTx.class);
+
+        JanusGraphElement indexableElement = spy(new StandardVertex(tx, 1l, ElementLifeCycle.New));
+        Property pk2 = indexableElement.property(key, value);
```
Refactored the code  :) and moved the mocked code into independent methods.

FlorianHockmann replied:
Thanks, that really improved readability a lot!
(One thing I'm noticing in your recent change are some empty lines at the end of some methods which shouldn't be there, but that's not that important.)

FlorianHockmann commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- janusgraph-test/src/test/java/org/janusgraph/graphdb/database/IndexSerializerTest.java  2
- janusgraph-core/src/main/java/org/janusgraph/graphdb/olap/job/IndexRepairJob.java  1
         

Complexity decreasing per file
==============================
+ janusgraph-core/src/main/java/org/janusgraph/diskstorage/keycolumnvalue/scan/StandardScannerExecutor.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/JanusGraph/janusgraph/pullRequest?prid=4609964&bid=15465456)

FlorianHockmann commented on janusgraph-core/src/main/java/org/janusgraph/diskstorage/keycolumnvalue/scan/StandardScannerExecutor.java:
```diff
@@ -146,9 +149,7 @@ public void run() {
-                        int retryCount = 0;
-                        while (!pullThreads[i].isFinished() && retryCount < TIMEOUT_MS / TIME_PER_TRY && qr == null) {
-                            retryCount ++;
+                        while (!pullThreads[i].isFinished() && qr == null) {
```
Didn't the `retryCount` ensure that this doesn't result in an infinite loop? You said in the issue that this fails sometimes for vertices with very many edges, but doesn't that mean now that the `TIMEOUT_MS` are ignored?

gguttikonda replied:
Hi @FlorianHockmann, in my case, code was reaching to max retryCount before indexing is complete. I tried increasing the TIMEOUT_MS, but it failed too for vertexes with large number of edges. Eg: a vertex having ~500K edges, CQL KeyIterator was taking a while to return the next element, and this loop was bailing before then. Because we moved the `finished=true` in DataPuller into finally block, this should never run into an infinite loop, at the same time giving enough time for while loop to continue looking for SliceResult until DataPuller says it's finished.

FlorianHockmann replied:
I tried to look into why this retry count was introduced and found the PR #1237 which fixed #1157.
@porunov Since you created #1157, could you take a look at this and check whether removing the timeout here could be problematic? Do you maybe have an easy way to ensure that it doesn't have negative effects on the issue you described in #1157?
Otherwise I'm OK with the explanation @gguttikonda gave and would accept this PR.

farodin91 replied:
@porunov Do you like to comment on this?

porunov replied:
Sorry I didn't get back on this earlier. I will review #2066 closer to weekends because this PR is obsolete now

porunov replied:
Replied here https://github.com/JanusGraph/janusgraph/pull/2066#pullrequestreview-391922488
This change doesn't reintroduce the issue. LGTM.

gguttikonda commented:
@farodin91 , @twilmes : can you please review this when you get a chance.

FlorianHockmann commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- janusgraph-core/src/main/java/org/janusgraph/graphdb/olap/job/IndexRepairJob.java  1
- janusgraph-test/src/test/java/org/janusgraph/graphdb/database/IndexSerializerTest.java  2
         

Complexity decreasing per file
==============================
+ janusgraph-core/src/main/java/org/janusgraph/diskstorage/keycolumnvalue/scan/StandardScannerExecutor.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/JanusGraph/janusgraph/pullRequest?prid=4609964&bid=15465456)

FlorianHockmann commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- janusgraph-test/src/test/java/org/janusgraph/graphdb/database/IndexSerializerTest.java  2
- janusgraph-core/src/main/java/org/janusgraph/graphdb/olap/job/IndexRepairJob.java  1
         

Complexity decreasing per file
==============================
+ janusgraph-core/src/main/java/org/janusgraph/diskstorage/keycolumnvalue/scan/StandardScannerExecutor.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/JanusGraph/janusgraph/pullRequest?prid=4609964&bid=15465456)

gguttikonda commented:
> LGTM! Thank you @gguttikonda.
> 
> Sorry, i forgot the PR fully.

Thank you @farodin91. I tried squashing all my commits into one and merged from master too. Please let me know if you have any feedback. If this PR is too lengthy, I am happy to create a new PR with all changes in one go.

farodin91 commented:
@gguttikonda I'm okay with squash in this PR or a different PR. It is your choice.

gguttikonda commented:
> @gguttikonda I'm okay with squash in this PR or a different PR. It is your choice.
Thanks @farodin91. I created a new PR ( https://github.com/JanusGraph/janusgraph/pull/2066 ) with single commit.
